include/extractor/guidance/validation_handler.hpp

```diff
@@ -177,6 +179,81 @@ class ValidationHandler final : public IntersectionHandler
             }
         }
     }
+
+    void checkForSharpDualCarriagewayUturns(const NodeID from_nid,
+                                            const EdgeID via_eid,
+                                            const Intersection &intersection) const
+    {
+        // Detect situations where dual-carriageway fans-in from two ways into one and
+        // the Uturn from a dual-carriageway onto the dual-carriageway should be possible.
+        //
+        // . .
+        //     > .
+        // . .
+        //
+
+        // Todo: are there dual-carriageway intersections of larger size?
+        if (intersection.size() != 3)
+        {
+            return;
+        }
+
+        const auto &via_road = intersection.getUTurnRoad();
+        const auto &leftmost_road = intersection.getLeftmostRoad();
+        const auto &rightmost_road = intersection.getRightmostRoad();
+
+        const auto &via_road_data = node_based_graph.GetEdgeData(via_eid);
+        const auto &leftmost_road_data = node_based_graph.GetEdgeData(leftmost_road.eid);
+        const auto &rightmost_road_data = node_based_graph.GetEdgeData(rightmost_road.eid);
+
+        if (via_road_data.name_id == EMPTY_NAMEID || leftmost_road_data.name_id == EMPTY_NAMEID ||
+            rightmost_road_data.name_id == EMPTY_NAMEID)
+        {
+            return;
+        }
+
+        const auto same_name_left =
+            !util::guidance::requiresNameAnnounced(via_road_data.name_id,
+                                                   leftmost_road_data.name_id,
+                                                   name_table,
+                                                   street_name_suffix_table);
+        const auto same_name_right =
+            !util::guidance::requiresNameAnnounced(via_road_data.name_id,
+                                                   rightmost_road_data.name_id,
+                                                   name_table,
+                                                   street_name_suffix_table);
+
+        if (!same_name_left || !same_name_right)
+        {
+            return;
+        }
+
+        if (!via_road.entry_allowed || !leftmost_road.entry_allowed ||
+            !rightmost_road.entry_allowed)
+        {
+            return;
+        }
+
+        const auto left_turn_angle = osrm::util::angularDeviation(0, leftmost_road.angle);
```

Comment:
angle is a bit confusing here. In addition: 2* NARROW could be a bit much (80 degrees).
OSRM will already try to correct for merging roads/offsets. So you should get away here with a far lower angle deviation.